stl/inc/format

```diff
@@ -445,103 +449,239 @@ _NODISCARD constexpr bool _Is_execution_charset_utf8() {
+
+        if constexpr (_Statically_Utf8) {
             return _Utf8_code_units_in_next_character(_First, _Last);
+        } else {
+            switch (this->_Cvt._Mbcurmax) {
+            default:
+                _STL_INTERNAL_CHECK(!"Bad number of encoding units for this code page");
+                [[fallthrough]];
+            case 1:
+                return 1; // all characters have only one code unit
+
+            case 2:
+                return this->_Double_byte_encoding_code_units_in_next_character(_First, _Last);
+
+            case 4: // Assume UTF-8 (as does _Mbrtowc)
+                return _Utf8_code_units_in_next_character(_First, _Last);
+            }
         }
     }
-}
 
-_NODISCARD inline int _Code_units_in_next_character(const wchar_t* _First, const wchar_t* _Last, const _Cvtvec&) {
-    // Returns a count of the number of code units that compose the first encoded character in
-    // [_First, _Last), or -1 if [_First, _Last) doesn't contain an entire encoded character or
-    // *_First is an unpaired surrogate.
-    _STL_INTERNAL_CHECK(_First < _Last);
+    _NODISCARD const char* _Find_encoded(const char* _First, const char* _Last, const char _Val) const {
+        // Returns the first occurrence of _Val as an encoded character (and not, for example, as a
+        // continuation byte) in [_First, _Last).
+        if constexpr (_Statically_Utf8) {
+            return _Find_unchecked(_First, _Last, _Val);
+        } else {
+            if (this->_Cvt._Mbcurmax == 1 || this->_Cvt._Mbcurmax == 4) {
+                // As above and in _Mbrtowc, assume 4-byte encodings are UTF-8
+                return _Find_unchecked(_First, _Last, _Val);
+            }
 
-    if (*_First < 0xD800u || *_First >= 0xE000u) {
-        return 1;
+            while (_First != _Last && *_First != _Val) {
+                const int _Units = _Units_in_next_character(_First, _Last);
+                if (_Units < 0) {
+                    _THROW(format_error("Invalid encoded character in format string."));
+                }
+                _First += _Units;
+            }
+
+            return _First;
+        }
     }
 
-    if (*_First >= 0xDC00u) { // unpaired low surrogate
-        return -1;
+    _NODISCARD int _Estimate_width(const char* const _Ptr, const int _Units) const {
+        // Return an estimate for the width of the character composed of _Units code units,
+        // whose first code unit is denoted by _Ptr.
+        if constexpr (_Statically_Utf8) {
+            return _Estimate_utf8_character_width(_Ptr, _Units);
+        } else {
+            if (this->_Cvt._Mbcurmax != 4) {
+                // not a Unicode encoding; estimate width == number of code units
+                return _Units;
+            }
+
+            // assume UTF-8
+            return _Estimate_utf8_character_width(_Ptr, _Units);
+        }
+    }
+};
+
+template <bool _Statically_Utf8>
+class _Fmt_codec<wchar_t, _Statically_Utf8> {
+private:

```

Comment:
Can we skip the private?

Reply:
GitHub really needs to add 🤦 to the list of "reaction" emoji.